docs: tutorial (#124)
electron/electronjs.org-new: docs/latest/tutorial/tutorial-scaffolding.md

of your project with the following contents and run `npm start` again:

```js
console.log(`Hello from Electron 👋`);
```

In your console you should read `Hello from Electron 👋`.

Congratulations, you have executed your first Electron application!

### Debugging from VS Code

If you want to debug your application using VS Code instead of using the terminal, do the
following:

1. Create a new folder `.vscode` in the root of your project.
1. Create a new file `launch.json` with the following content:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Debug Electron's Main Process",
      "type": "node",
      "request": "launch",
      "cwd": "${workspaceFolder}",
      "runtimeExecutable": "${workspaceFolder}/node_modules/.bin/electron",
      "windows": {
        "runtimeExecutable": "${workspaceFolder}/node_modules/.bin/electron.cmd"
      },
      "args": ["."],
      "outputCapture": "std"
    }
  ]
}
```

3. The option `Debug Electron's Main Process` will appear when you select `Run and Debug`
   from the sidebar, allowing you to set breakpoints and inspect all the variables among
   other things.

> ![Screenshot of VS Code debugging the code so far](/assets/img/vscode-debugger.png)

Reviewer: I mean the entire VSCode debugger part

> ![Screenshot of VS Code debugging the code so far](/assets/img/vscode-debugger.png)

Author: I think the whole section would look weird as an admonition unless we split it into another file. The reality is that the vast majority of Electron developers are using VS Code so I'd just make this information as visible as possible (and if we ever create custom boilerplate I'd add those files as well).

> ![Screenshot of VS Code debugging the code so far](/assets/img/vscode-debugger.png)

Author: I've grouped all the VS Code debugging and put it into the scaffolding part

:::

The interactive `init` command will prompt you to set some fields in your config.
There are a few rules to follow for the purposes of this tutorial:

- `entry point` should be `main.js`.
- `author` and `description` can be any value, but are necessary for
  [app packaging][application distribution].

Your `package.json` file should look something like this:

```json title='package.json'
{
  "name": "my-electron-app",
  "version": "1.0.0",
  "description": "Hello World!",
  "main": "main.js",
  "author": "Jane Doe",
  "license": "MIT"
}
```

Then, install the `electron` package into your app's `devDependencies`.

```sh npm2yarn
npm install --save-dev electron
```

Finally, you want to be able to execute Electron. In the [`scripts`][package-scripts]
field of your `package.json` config, add a `start` command like so:

```json {8-10} title='package.json'
{
  "name": "my-electron-app",
  "version": "1.0.0",
  "description": "Hello World!",
  "main": "main.js",
  "author": "Jane Doe",
  "license": "MIT",
  "scripts": {
    "start": "electron ."
  }
}
```

This `start` command will let you open your app in development mode.

```sh npm2yarn
npm start
```

> :::tip

Author: ```suggestion
:::info
```

This script tells Electron to run on your project's root folder. At this stage,
your app will immediately throw an error telling you

> "Error launching app"

This is expected as we have not writen any code yet!
:::

## Adding the entry point

The entry point of any Electron application is its `main` script. This script controls the
**main process**, which runs in a full Node.js environment and is responsible for
controlling your app's lifecycle, displaying native interfaces, performing privileged
operations, and managing renderer processes (more on that later).

During execution, Electron will look for this script in the [`main`][package-json-main]
field of the app's `package.json` config, which you should have configured in the previous
step.

To initialize the `main` script, create a file named `main.js` in the root folder
of your project with the following contents and run `npm start` again:

```js title='main.js'
console.log(`Hello from Electron 👋`);
```

In your console you should read `Hello from Electron 👋`.

> Congratulations, you have executed your first Electron application!

Author: But it's true! Need to keep them motivated :D

:::tip
If you haven't yet, please read [Electron's process model][process-model] to better understand
how all the pieces work together.
:::

> ## Adding some UI

Reviewer: ```suggestion
## Creating a window
```

will be loaded into it. In Electron, each window displays web contents that can be loaded
from either from a local HTML file or a remote URL.

For this tutorial, you will be doing the former. Create an `index.html` file in the root
folder of your project:

```html title='index.html'
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <!-- https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP -->
    <meta
      http-equiv="Content-Security-Policy"
      content="default-src 'self'; script-src 'self'"
    />
    <meta
      http-equiv="X-Content-Security-Policy"
      content="default-src 'self'; script-src 'self'"
    />
    <title>Hello from Electron renderer!</title>
  </head>
  <body>
    <h1>Hello from Electron renderer!</h1>
    <p>👋</p>
  </body>
</html>
```

Now that you have a web page, load it into an application window. To do so, paste
the following code in your `main.js` file so we can disect it line by line:

```js title='main.js'
const { app, BrowserWindow } = require('electron')

const createWindow = () => {
  const win = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {

>       sandbox: true,

Reviewer: I think we should align with `electron-quick-start` for props here. Not sure if we should be including `sandbox` without greater alignment around `sandbox: true` across the docs, or at least more explanation on what the prop does here and linking to the Sandboxing doc.

Also unrelated but there's a trailing comma here but none in the next instance of this snippet.

Author: weird that the comma isn't failing in the pr to e/e.

I'd rather use `sandbox` in the tutorial. That's what we want people to use and the default in this cases should be `true` soon-ish when not specified.

Author: ```suggestion
      sandbox: true
```

    }
  })

  win.loadFile('index.html')
}

app.whenReady().then(() => {
  createWindow()
})
```

In the first line we are `require`ing 2 Electron modules:

- The [`app`][app] module, which controls your application's event lifecycle.
- The [`BrowserWindow`][browser-window] module, which creates and manages application
  windows.

```js
const { app, BrowserWindow } = require('electron')
```

:::warning

> ESM (i.e.: use `import` to load a module) is currently not supported in Electron directly.

Reviewer: ```suggestion
[ECMAScript modules](https://nodejs.org/api/esm.html) (i.e. using `import` to load a module) is currently not supported in Electron directly.
```

You can find more information about the state and chanllenges in
[this issue](https://github.com/electron/electron/issues/21457).
:::

Then, `createWindow()` function loads `index.html` into a new `BrowserWindow`
instance with a series of options:

```js
const createWindow = () => {
  const win = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      sandbox: true
    }
  })

  win.loadFile('index.html')
}
```

:::tip

> You can find all the options available in the [`BrowserWindow` page][browser-window].

Reviewer: ```suggestion
You can find all the constructor options available in the [`BrowserWindow` API docs][browser-window].
```

:::

In Electron, browser windows can only be created after the `app` module's
[`ready`][app-ready] event is fired. You can wait for this event by using the
[`app.whenReady()`][app-when-ready] API. Call `createWindow()` after `whenReady()`
resolves its Promise. This is done in the last 3 lines of the code:

```js
app.whenReady().then(() => {
  createWindow()
})
```

> At this point, your Electron application should successfully open a window that displays your
> web page!

Reviewer: I realize now that we don't really introduce the concept of the renderer process at this point:

```suggestion
At this point, your Electron application should successfully open a window that displays your
web page! Each window created by your app will run in a separate process called a
**renderer**.
```

You have now code being executed in the main and renderer process.
The easiest way to debug the code in your renderer process is by using the built-in developer
tools. You can invoke them with a shortcut (`Ctrl/Command+Shift+I`) or programmatically

> via `mainWindow.webContents.openDevTools()` in the main process. If you want to use VS Code,

Reviewer: ```suggestion
via `win.webContents.openDevTools()` in the main process. If you want to use VS Code,
```
Aligning the window reference variable name here

keep reading!

The renderer has access to the web APIs. That means you can use the same JavaScript APIs and
tooling you use for typical front-end development, such as using [`webpack`][webpack] to bundle
and minify your code or [React][react] to manage your user interfaces.

:::tip
You can find more information about this subject in the [Boilerplates and CLIs].
:::

> ### Debugging from VS Code

Reviewer: Is this meant to be an H3? I feel it'd fit better as an H2.

Reviewer: ```suggestion
### Optional: Debugging from VS Code
```

Author: ```suggestion
## Optional: Debugging from VS Code
```

keep in mind there are 2 processes to attach at this moment: the main and renderer processes.

We are going to create a configuration file that allow us to debug both at the same time:

1. Create a new folder `.vscode` in the root of your project.
1. Create a new file `launch.json` with the following content:

```json title='.vscode/launch.json'
{
  "version": "0.2.0",
  "compounds": [
    {
      "name": "Main + renderer",
      "configurations": ["Main", "Renderer"],
      "stopAll": true
    }
  ],
  "configurations": [
    {
      "name": "Renderer",
      "port": 9222,
      "request": "attach",
      "type": "pwa-chrome",
      "webRoot": "${workspaceFolder}"
    },
    {
      "name": "Main",
      "type": "pwa-node",
      "request": "launch",
      "cwd": "${workspaceFolder}",
      "runtimeExecutable": "${workspaceFolder}/node_modules/.bin/electron",
      "windows": {
        "runtimeExecutable": "${workspaceFolder}/node_modules/.bin/electron.cmd"
      },
      "args": [".", "--remote-debugging-port=9222"],
      "outputCapture": "std",
      "console": "integratedTerminal"
    }
  ]
}
```

<!-- markdownlint-disable MD029 -->

> 3. The option `Main + renderer` will appear when you select `Run and Debug`

Reviewer: ```suggestion
1. The option `Main + renderer` will appear when you select `Run and Debug`
```

Author: It has to be a 3, otherwise the numbering starts again at 1. That's why I have disabled the markdownlint rule

Reviewer: Oh I see! I missed that.

from the sidebar, allowing you to set breakpoints and inspect all the variables among
other things in both the main and renderer processes.
<!-- markdownlint-enable MD029 -->

> What we have done in the `lauch.json` file is to create 3 configurations:

Reviewer: ```suggestion
What we have done in the `launch.json` file is to create 3 configurations:
```

> - `Main` is use to start the main process and also expose the port 9222 for remote debugging

Reviewer: ```suggestion
- `Main` is used to start the main process and also expose port 9222 for remote debugging
```

  (`--remote-debugging-port=9222`). This is the port that we will use to attach the debugger
  for the `Renderer`. Because the main process is a Node.js process, the type is set to
  `pwa-node` (`pwa-` is the prefix that tells VS Code to use the latest JavaScript debugger).
- `Renderer` is used to debug the renderer process. It tries to connect to the port 9222.
  Because the renderer is a web process, the debugger that we use is `pwa-chrome`.
- `Main + renderer` is a [compound task] that executes the previous ones simulatenously.

:::caution
Because we are attaching to a process that exists already in the `Renderer` configuration

> (`"request": "attach"`) there is a high chance that you will not be able to set a breakpoint

Reviewer: I think we should explain the `request: attach` thing above. I'm not too familiar with it myself. Does it mean that we're attaching to the existing remote debugging port at 9222 created when running the "Main" configuration?

Does this also work with multiple renderer processes?

Author: > Does it mean that we're attaching to the existing remote debugging port at 9222 created when running the "Main" configuration?

Yes, haven't tried the multiple renderer process.
I'll make it clearer.

Author: I've updated the `request: "attach"` in case you want to take another look

in the first lines of your code as the debugger will not have had enough time to connect before
they are being executed.

> You can workaround this by refreshing the page or setting a timeout before executing the code

Reviewer: ```suggestion
You can work around this by refreshing the page or setting a timeout before executing the code
```

in development mode.
:::

> :::tip Further reading

Reviewer: Should we add titles to all the other admonitions as well? :)

Author: There are some that make sense, others with the default "Note" or "Tip" I think are good. I'll review them just in case.
